### PR TITLE
Adapt subscribed api

### PR DESCRIPTION
**Pull request by muralov**
Adapt /subscribed endpoint to v1alhpa2 CRD

**Related issue(s)**
#[15624](https://github.com/kyma-project/kyma/issues/15624)

### CODE REVIEW

File: components/event-publisher-proxy/pkg/subscribed/helpers.go [hunk 1]
```diff
@@ -72,10 +91,49 @@ func AddUniqueEventsToResult(eventsSubSet []Event, uniqEvents map[Event]bool) ma
...
+				event := buildEvent(eventType)
+				events = append(events, event)
+			}
+		}
+	}
+	return events
+}
+
+// it receives event and type version, e.g. order.created.v1 and returns {Name: order.created, Version: v1}
+func buildEvent(eventTypeAndVersion string) Event {
```
Comment (raypinto) [inline components/event-publisher-proxy/pkg/subscribed/helpers.go:122]: (nit) can you write a unit test for this function as well. I'm aware that its called in the previous function and its unit tested, but would be nice to unit test the smallest block as well

File: components/event-publisher-proxy/pkg/subscribed/helpers.go [hunk 2]
```diff
@@ -72,10 +91,49 @@ func AddUniqueEventsToResult(eventsSubSet []Event, uniqEvents map[Event]bool) ma
...
 
 // FilterEventTypeVersions returns a slice of Events:
+// 1. if the eventType matches the format for standard type matching: <event-name>.<version>
+// E.g. order.created.v0
+// 2. if the eventType matches the format for exact type matching: <eventTypePrefix><appName>.<event-name>.<version>
+// E.g. sap.external.custom.exampleapp.order.created.v0
+func FilterEventTypeVersions(appName string, subscription *eventingv1alpha2.Subscription) []Event {
+	events := make([]Event, 0)
+	for _, eventType := range subscription.Spec.Types {
+		if subscription.Spec.TypeMatching == eventingv1alpha2.TypeMatchingExact {
```
Comment (raypinto) [inline components/event-publisher-proxy/pkg/subscribed/helpers.go:101]: I have updated the pseudocode [here](https://github.com/kyma-project/kyma/issues/15624#issuecomment-1342224345) to have one more check. Could you please implement it as well

File: components/event-publisher-proxy/pkg/subscribed/helpers.go [hunk 3]
```diff
@@ -72,10 +91,49 @@ func AddUniqueEventsToResult(eventsSubSet []Event, uniqEvents map[Event]bool) ma
...
+			// in case of type matching exact, we have app name as a part of event type
+			appNameWithDots := "." + appName + "."
+			if strings.Contains(eventType, appNameWithDots) {
+				eventTypeVersionArr := strings.Split(eventType, appNameWithDots)
+				eventTypeVersion := eventTypeVersionArr[len(eventTypeVersionArr)-1]
```
Comment (raypinto) [inline components/event-publisher-proxy/pkg/subscribed/helpers.go:106]: Maybe optimize these two lines into one as:
```
eventTypeVersion  := strings.SplitAfter(eventType, appNameWithDots)[1]
```

**Review by raypinto — COMMENTED** [review]
Also bump the EPP image

File: components/event-publisher-proxy/pkg/subscribed/helpers_test.go [hunk 4]
```diff
@@ -1,13 +1,117 @@
...
+				Spec: eventingv1alpha2.SubscriptionSpec{
+					Source:       "/default/sap.kyma/tunas-develop",
+					TypeMatching: eventingv1alpha2.TypeMatchingExact,
+					Types: []string{
+						"foo.foovarkes.custom.foovarkes.order.created.v1",
+					},
+				},
+			},
+			expectedEvents: []Event{
+				NewEvent("order.created", "v1"),
```
Comment (raypinto) [inline components/event-publisher-proxy/pkg/subscribed/helpers_test.go:96]: as per the updated pseudocode this should return empty Event array

File: components/event-publisher-proxy/pkg/subscribed/helpers.go (hunk 3 above)
Comment (raypinto) [reply components/event-publisher-proxy/pkg/subscribed/helpers.go:106]: I haven't tested this. Just a hint! Please try it out and see if there are edge cases

Comment (raypinto) [reply components/event-publisher-proxy/pkg/subscribed/helpers.go:106]: I think it wouldn't work if the appName is `kyma` and the prefix contains `kyma`. Maybe a testcase to add :)

File: components/event-publisher-proxy/pkg/subscribed/helpers.go [hunk 5]
```diff
@@ -95,15 +95,15 @@ func AddUniqueEventsToResult(eventsSubSet []Event, uniqEvents map[Event]bool) ma
 // E.g. order.created.v0
 // 2. if the eventType matches the format for exact type matching: <eventTypePrefix><appName>.<event-name>.<version>
 // E.g. sap.external.custom.exampleapp.order.created.v0
```
Comment (raypinto) [inline components/event-publisher-proxy/pkg/subscribed/helpers.go:97]: (nit) this comment is outdated

File: components/event-publisher-proxy/pkg/subscribed/helpers_test.go [hunk 6]
```diff
@@ -1,13 +1,154 @@
...
+					Source: "diff-source",
+					Types: []string{
+						"order.created.v1",
+						"order.created.v2",
+					},
+				},
+			},
+			expectedEvents: []Event{},
+		}, {
+			name:    "should return event types w/o prefix for exact external event types",
```
Comment (raypinto) [inline components/event-publisher-proxy/pkg/subscribed/helpers_test.go:53]: the title does not sound right now. something to change here

File: components/event-publisher-proxy/pkg/subscribed/helpers_test.go [hunk 7]
```diff
@@ -1,13 +1,154 @@
...
+					TypeMatching: eventingv1alpha2.TypeMatchingExact,
+					Types: []string{
+						"sap.kyma.custom.difffoovarkes.order.created.v1",
+						"sap.kyma.custom.difffoovarkes.order.created.v2",
+					},
+				},
+			},
+			expectedEvents: []Event{},
+		}, {
+			name:    "should return event type only with 'sap.kyma.custom' prefix",
```
Comment (raypinto) [inline components/event-publisher-proxy/pkg/subscribed/helpers_test.go:84]: can you update this to:
```suggestion
			name:    "should return event type only with 'sap.kyma.custom' prefix and appname",
```

File: components/event-publisher-proxy/pkg/subscribed/helpers_test.go [hunk 8]
```diff
@@ -1,13 +1,154 @@
...
+			appName: "foovarkes",
+			subscription: &eventingv1alpha2.Subscription{
+				Spec: eventingv1alpha2.SubscriptionSpec{
+					Source:       "/default/sap.kyma/tunas-develop",
+					TypeMatching: eventingv1alpha2.TypeMatchingExact,
+					Types: []string{
+						"foo.prefix.custom.foovarkes.order.created.v1",
+						"sap.kyma.custom.foovarkes.order.created.v2",
```
Comment (raypinto) [inline components/event-publisher-proxy/pkg/subscribed/helpers_test.go:92]: and add an additional types here:
"foo.prefix.custom.diffvarkes.order.created.v1",
"sap.kyma.custom.diffvarkes.order.created.v2",

Comment (raypinto) [reply components/event-publisher-proxy/pkg/subscribed/helpers_test.go:92]: or add another test case like this

File: components/eventing-controller/pkg/deployment/publisher_deployment.go [hunk 9]
```diff
@@ -316,6 +316,10 @@ func getBEBEnvVars(publisherConfig env.PublisherConfig) []v1.EnvVar {
 			Name:  "BEB_NAMESPACE",
 			Value: fmt.Sprintf("%s$(BEB_NAMESPACE_VALUE)", bebNamespacePrefix),
 		},
+		{
+			Name:  "ENABLE_NEW_CRD_VERSION",
+			Value: strconv.FormatBool(publisherConfig.EnableNewCRDVersion),
+		},
```
Comment (raypinto) [inline components/eventing-controller/pkg/deployment/publisher_deployment.go:322]: why do you need this?

File: components/event-publisher-proxy/pkg/subscribed/helpers_test.go [hunk 10]
```diff
@@ -90,6 +90,7 @@ func TestFilterEventTypeVersions(t *testing.T) {
 					Types: []string{
 						"foo.prefix.custom.foovarkes.order.created.v1",
 						"sap.kyma.custom.foovarkes.order.created.v2",
+						"sap.kyma.custom.diffvarkes.order.created.v2",
 					},
```
[the remark this answers is not in the document]
Comment (muralov) [reply components/event-publisher-proxy/pkg/subscribed/helpers_test.go:95]: `"foo.prefix.custom.foovarkes.order.created.v1"` is with wrong prefix, but with the right appName

File: components/eventing-controller/pkg/deployment/publisher_deployment.go (hunk 9 above)
Comment (muralov) [reply components/eventing-controller/pkg/deployment/publisher_deployment.go:322]: I thought event if Event Mesh is enabled, the `/subscribed`endpoint should be callable.

File: components/eventing-controller/pkg/env/backend_config.go [hunk 11]
```diff
@@ -36,8 +36,9 @@ type PublisherConfig struct {
 	LimitsMemory      string `envconfig:"PUBLISHER_LIMITS_MEMORY" default:"128Mi"`
 	PriorityClassName string `envconfig:"PUBLISHER_PRIORITY_CLASS_NAME" default:""`
 	// publisher takes the controller values
-	AppLogFormat string `envconfig:"APP_LOG_FORMAT" default:"json"`
-	AppLogLevel  string `envconfig:"APP_LOG_LEVEL" default:"info"`
+	AppLogFormat        string `envconfig:"APP_LOG_FORMAT" default:"json"`
+	AppLogLevel         string `envconfig:"APP_LOG_LEVEL" default:"info"`
+	EnableNewCRDVersion bool   `envconfig:"ENABLE_NEW_CRD_VERSION" default:"false"`
```
[the remark this answers is not in the document]
Comment (muralov) [reply components/eventing-controller/pkg/env/backend_config.go:41]: Please look at the comment above regarding having ENABLE_NEW_CRD_VERSION flag for the Event Mesh case.

File: components/eventing-controller/pkg/deployment/publisher_deployment.go (hunk 9 above)
Comment (raypinto) [reply components/eventing-controller/pkg/deployment/publisher_deployment.go:322]: how does the `/subscribed` endpoint depend on the backend enabled?

File: components/event-publisher-proxy/pkg/subscribed/helpers.go (hunk 1 above)
Comment (muralov) [reply components/event-publisher-proxy/pkg/subscribed/helpers.go:125]: done

File: components/event-publisher-proxy/pkg/subscribed/helpers.go [hunk 12]
```diff
@@ -72,10 +91,51 @@ func AddUniqueEventsToResult(eventsSubSet []Event, uniqEvents map[Event]bool) ma
 }
 
 // FilterEventTypeVersions returns a slice of Events:
+// 1. if the eventType matches the format for standard type matching: <event-name>.<version>
+// and source is equal to appName.
+// E.g. order.created.v0
+// 2. if the eventType matches the format for exact type matching: <eventTypePrefix>.<appName>.<event-name>.<version>
+// and eventTypePrefix and appName together is a prefix.
+// E.g. sap.kyma.custom.exampleapp.order.created.v0
```
[the remark this answers is not in the document]
Comment (muralov) [reply components/event-publisher-proxy/pkg/subscribed/helpers.go]: done

File: components/event-publisher-proxy/pkg/subscribed/helpers_test.go [hunk 13]
```diff
@@ -1,13 +1,155 @@
...
+					Source: "diff-source",
+					Types: []string{
+						"order.created.v1",
+						"order.created.v2",
+					},
+				},
+			},
+			expectedEvents: []Event{},
+		}, {
+			name:    "should return event types if event type has prefix and app name as prefix",
```
[the remark this answers is not in the document]
Comment (muralov) [reply components/event-publisher-proxy/pkg/subscribed/helpers_test.go]: done

**Review by raypinto — DISMISSED** [review]

**Review by raypinto — APPROVED** [review]

**Pull request merged**